enhancement_template: add API extension sections (#933)
openshift/enhancements: guidelines/enhancement_template.md

approvers:
  - TBD
  - "@oscardoe"
api-approvers:

>   - TBD

Author: Would like to link to the list of possible approvers. https://github.com/openshift/api/blob/master/OWNERS is not that. An OWNERS_ALIASES file might work as in upstream.

>   - TBD

Reviewer: We're planning to change the way we list reviewers generally, to include the scope of review being requested and other details to help the final approver ensure that all areas are covered. I think we can address the question of how to find people for different types of reviews as part of that change.

Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.

> ### API Extensions

Reviewer 2: Q: why is this section so far from the Impact of API Extensions section? Is there a reason to seperate them? Does doing so impact readability?

Author: the impact is in the details part, the listing is here, for easier reviewing.

> ### API Extension Impact

Author: added

> - name the API extensions (CRDs, webhooks, aggregated API servers) this enhancement adds or modifies

Reviewer 2: ```suggestion
- Name the API extensions (CRDs, webhooks, aggregated API servers) this enhancement adds or modifies
```
Capitalize throughout

Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.

### API Extensions

- name the API extensions (CRDs, webhooks, aggregated API servers) this enhancement 

> - adds or modifies

Reviewer 3: part of previous sentence?

- does this enhancement modify the behaviour of existing resources and if yes how

  Examples:
    - adds a finalizer to namespaces. Namespace cannot be deleted without our controller running.
    - restricts the label format for objects to X.

>     - defaults field Y on object kind Z.

Reviewer 3: would you also expect new resource types, and/or new fields being added to existing resource types, to be described here?

(I would, though they also tend to end up in the implementation details section currently)

Reviewer 3: hm, I don't see why "who owns the api" vs "who is writing the EP" matters?  That may affect who needs to review it, but not what the design needs to cover.

what matters is the kind of change you are proposing to the product, and the effects it may have that need to be considered.

For which it seems like "api" changes need to be broken down into a few categories, e.g.:

1) new api type
2) updating existing api type (e.g. new fields)
3) adding a webhook (not sure if conversion vs admission need to be divided)
4) adding/changing validations on an apitype (adding is usually not allowed, of course)
5) adding/changing rbac associated w/ an api type

i'm not saying you need to cover all of them in your changes here, but i think it's important we make it clear which kinds of changes these new sections are applicable to.

Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.

### API Extensions

- name the API extensions (CRDs, admission and conversion webhooks, aggregated API servers) 
  this enhancement adds or modifies
- does this enhancement modify the behaviour of existing resources, especially those owned 

>   by other parties than the authoring team (including upstream resources), and if yes how

Reviewer 2: ```suggestion
  by other parties than the authoring team (including upstream resources), and, if yes, how?
```

>   Examples:

Reviewer 2: Do you want to add a specific example of something owned by someone else and how you'd like to see that described?

Author: the examples below are like that. A finalizer on namespaces modifies a kube object. A label restriction modifies another object normal behaviour.

Reviewer 2: Q: if it's owned by someone else do you want them to explicitly say it like you requested like in failure modes below? Should they be added as reviewers?

Author: I would leave that to the API review to judge. Some APIs have owners that can be added. If the modification is more generic, i.e. to multiple resources, it's not that clear.

Reviewer 2: lgtm I just don't have perms to resolve my comment :)

Include a story on how this proposal will be operationalized:  lifecycled, monitored and remediated at scale.

### API Extensions

API Extensions are CRDs, admission and conversion webhooks, aggregated API servers,

> finalizers, i.e. those mechanisms that change the OCP API surface and behaviour.

Author: CRDs certainly have less operation impact. But listing them is a good idea IMO.

I fear a little that we have to try to keep the balance. We can add more and more to stop certain common mistakes from happening. But the risk is that ppl neglect certain sections because there are just too many. I notice this a lot with the risk, test and version skew section at the bottom. Often they are not seriously filled in.

> finalizers, i.e. those mechanisms that change the OCP API surface and behaviour.

Reviewer 3: can we give more guidance about when those sections are applicable?  my fear is that someone adding a new CRD is going to have no idea what to put in the "SLI" section (i don't really know what we'd expect them to put there)

- Will any other components on the node change? For example, changes to CSI, CRI
  or CNI may require updating that component before the kubelet.

### Impact of API Extensions

Describe the API extensions here in detail, especially their impact on a cluster:
 
- what are the SLIs (Service Level Indicators) an administrator or support can use to 
- determine the health of the API extensions

  Examples: metrics, alerts, operator conditions
- which impact do these API extensions have on existing SLIs (e.g. scalability, API throughput,
  API availability)

  Examples:
    - adds 1s to every pod update in the system
    - fails creation of ConfigMap in the system when the webhook is not availiable
    - adds a dependency on the SDN service network for all resources, risking API availability in case of SDN issues
- how is the upper impact to be measured and when (e.g. every release by QE, or automatically in CI)
  and by whom (e.g. perf team; name the responsible person and let them review this enhancement)

#### Failure Modes

> - describe the possible failure modes of the API extensions

Reviewer 2: Throughout can we follow the existing text formatting of beginning lines/sentences with a capital letter? Makes template consistent and more readable.

Author: done

- Will any other components on the node change? For example, changes to CSI, CRI
  or CNI may require updating that component before the kubelet.

### Impact of API Extensions

> Describe the API extensions here in detail, especially their impact on a cluster:

Reviewer 3: this seems very focused on the case where a new api resource type is being added, but doesn't cover the case where someone is trying to extend an existing api with new fields or behavior.

how are "api extensions" defined?

1) new resource type?
2) new field on a resource type?
3) new webhook on an existing type (or being added as part of adding a new resource type)?

Reviewer 3: actually i guess it's focused on the case where a new webhook or finalizer is being added?

Reviewer 4: As stated above `API Extensions` is everything impacting APIs: webhooks, finalizers, CRDs. 
The effect of a new API field being added should be describe throughout the enhancement discussing that kind of extension of the existing API, no?

 
- what are the SLIs (Service Level Indicators) an administrator or support can use to 

> - determine the health of the API extensions

Reviewer 3: this could use a more explicit example for people to follow, e.g. "when api extension $foo was added, the following SLI was introduced...."

if i'm adding a new config resource for builds(for example), i am not clear what it would mean to add a new SLI for that resource, or if such a thing is even really applicable to that api change.

- Will any other components on the node change? For example, changes to CSI, CRI
  or CNI may require updating that component before the kubelet.

### Impact of API Extensions

Describe the API extensions here in detail, especially their impact on a cluster:

- for conversion/admission webhooks and aggregated apiservers: what are the SLIs (Service Level Indicators)
  an administrator or support can use to determine the health of the API extensions

  Examples (metrics, alerts, operator conditions)
  - authentication-operator condition `APIServerDegraded=False`
  - authentication-operator condition `APIServerAvailable=True`
  - openshift-authentication/oauth-apiserver deployment and pods health

- which impact do these API extensions have on existing SLIs (e.g. scalability, API throughput,
  API availability)

  Examples:
  - adds 1s to every pod update in the system
  - fails creation of ConfigMap in the system when the webhook is not available
  - adds a dependency on the SDN service network for all resources, risking API availability in case of SDN issues

> - how is the upper impact to be measured and when (e.g. every release by QE, or automatically in CI)

Author: the impact on the SLI, shown above. Will clarify.

  and by whom (e.g. perf team; name the responsible person and let them review this enhancement)

#### Failure Modes

- describe the possible failure modes of the API extensions

> - describe how a failure or behaviour of the extension will impact the overall cluster health

Author: there might be behaviours, like slowing down requests or adding finalizers, that influence cluster health.

  (e.g. which kube-controller-manager functionality will stop working), especially regarding
  stability, availability, performance and security.

> - describe which teams are probably called out in case of escalation with one of the failure modes

Reviewer 2: ```suggestion
- Describe which OCP teams are likely to be called upon in case of escalation with one of the failure modes
```

Reviewer 2: I really like this btw

Author: done

>   and add them as reviewer to the enhancement.

Reviewer 2: ```suggestion
  and add them as reviewers to this enhancement.
```

  - operator X will degrade with message "Failed to launch webhook server" and reason "WehhookServerFailed"
  - the metric `webhook_admission_duration_seconds("openpolicyagent-admission", "mutating", "put", "false")`
    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.

- disable the API extension (e.g. remove MutatingWebhookConfiguration `xyz`, remove APIService `foo`)
    
    - which consequences does it have on the cluster health?
    
      Examples:
      - garbage collection in kube-controller-manager will stop working.
      - quota will be wrongly computed.

    - which consequences does it have on existing, running workloads?
  
      Examples:
      - new namespaces won't get the finalizer "xyz" and hence might leak resource X
        when deleted
      - SDN pod-to-pod routing will stop updating, potentially breaking pod-to-pod
        communication after some minutes.
      
    - which consequences does it have for newly created workloads?
  
      Examples:
      - new pods in namespace with Istio support will not get sidecars injected, breaking
        their networking

>     - does functionality fail gracefully and will resume work when re-enabled without risking

Author: done
